misc/python/materialize/cli/mzcompose.py

```diff
@@ -177,142 +189,251 @@ def gen_shortcuts(repo: mzbuild.Repository) -> int:
+UnpauseCommand = DockerComposeCommand("unpause", "unpause services")
+UpCommand = DockerComposeCommand("up", "create and start containers", acquire_deps=True)
+VersionCommand = DockerComposeCommand("version", "show version information and quit")
+
+
+# An `ArgumentParser` that attaches unknown arguments under the `unknown_args`
+# field rather than throwing an error.
+class ArgumentParser(argparse.ArgumentParser):
     def parse_known_args(
         self,
         args: Optional[Sequence[Text]] = None,
         namespace: Optional[argparse.Namespace] = None,
     ) -> Tuple[argparse.Namespace, List[str]]:
-        ns = argparse.Namespace()
-        try:
-            (pargs, unknown_args) = super().parse_known_args(args, namespace=ns)
-            if pargs.mz_build_mode not in ["dev", "release"]:
-                raise errors.BadSpec(
-                    f'unknown build mode {pargs.mz_build_mode!r} (expected "dev" or "release")'
-                )
-            return (pargs, unknown_args)
-        except ValueError:
-            return (ns, [])
-
-    def print_help(self, file: Optional[IO[str]] = None) -> None:
-        print(
-            f"""mzcompose orchestrates services defined in mzcompose.yml.
-It wraps Docker Compose to add some Materialize-specific features.
-
-Usage:
-  mzcompose [options] COMMAND [ARGS...]
-
-Options:
-  --mz-build-mode <dev|release>
-                     Specify the Cargo profile to use when compiling Rust
-                     crates (default: release)
-  --mz-coverage      Emit code coverage reports to the "coverage" directory
-  --mz-find DIR      Use the mzcompose.yml file from DIR, rather than the
-                     current directory
-  --mz-quiet         Suppress Materialize-specific informational messages
-  --verbose          Show more output
-
-Commands:
-  build              Build or rebuild services
-  down               Stop and remove containers, networks, images, and volumes
-  gen-shortcuts      Generate shortcut `mzcompose` shell scripts in mzcompose directories
-  lint               Surface common errors in compositions
-  list-compositions  List the directories that contain compositions
-  list-ports SERVICE List any ports exposed by SERVICE
-  list-workflows     List the workflows in the composition
-  ps                 List containers
-  run                Run a one-off command or workflow
-  start              Start services
-  stop               Stop services
-  up                 Create and start containers
-  web SERVICE        Open SERVICE's URL in a web browser
-
-These are only the most common commands. There are additional Docker Compose
-commands that are also supported. Consult `docker-compose help` for the full
-set.
-
-For additional details on mzcompose, consult doc/developer/mzbuild.md.""",
-            file=file,
-        )
+        args, unknown_args = super().parse_known_args(args, namespace)
+        setattr(args, "unknown_args", unknown_args)
```

Comment:
the argparse api is weird, do you not set it on the `Namespace` here instead?

Reply:
also, instead of this, it may be better to use `nargs=argparse.REMAINDER`: https://stackoverflow.com/questions/22850332/getting-the-remaining-arguments-in-argparse/46250042

It doesn't look like any command distinguishes between `unknown_args` and `unknown_subargs`, so it might be cleaner, however, its been a while since I've used that, so ymmv

Reply:
Heh, I went down that road. `argparse.REMAINDER` was deprecated in Python 3.9 because it has unfixable bugs: http://bugs.python.org/issue17050. It also doesn't distinguish between unknown args before the subcommand and unknown args after the subcommand, which is extremely important for properly forwarding args to `docker-compose`.

Reply:
I added that as a comment though! 

> the argparse api is weird, do you not set it on the `Namespace` here instead?

You _do_, it's just that `args` becomes the namespace. I refactored to make it clearer. (Sorry, I copied this out of the Python standard library default impl of `parse_known_args`, which had that confusing naming bit.)